--- a/development/src/GXDLMS.cpp
+++ b/development/src/GXDLMS.cpp
@@ -881,7 +881,9 @@ int CGXDLMS::GetLNPdu(
         }
 
         // Add attribute descriptor.
-        reply.Set(p.GetAttributeDescriptor());
+        if (!((p.GetCommand() == DLMS_COMMAND_SET_REQUEST) && (p.GetRequestType() == 3)))
+            reply.Set(p.GetAttributeDescriptor());
+
         // If multiple blocks.
         if (p.IsMultipleBlocks() && (p.GetSettings()->GetNegotiatedConformance() & DLMS_CONFORMANCE_GENERAL_BLOCK_TRANSFER) == 0)
         {
@@ -933,7 +935,7 @@ int CGXDLMS::GetLNPdu(
                 len -= GXHelpers::GetObjectCountSizeInBytes(len);
             }
             GXHelpers::SetObjectCount(len, reply);
-            reply.Set(p.GetData(), 0, len);
+            reply.Set(p.GetData(), p.GetData()->GetPosition(), len);
         }
         // Add data that fits to one block.
         if (len == 0)
@@ -4159,4 +4161,4 @@ int CGXDLMS::HandleConfirmedServiceError
         return service << 16 | type << 8 | ch;
     }
     return 0;
-}
\ No newline at end of file
+}
